src/compute/src/extensions/collection.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use differential_dataflow::difference::Semigroup;
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::operators::arrange::{Arranged, TraceAgent};
+use differential_dataflow::trace::{Batch, Trace, TraceReader};
+use differential_dataflow::{Collection, Data, ExchangeData, Hashable};
+use timely::dataflow::Scope;
+
+use crate::extensions::operator::{ArrangementSize, MzArrange};
+
+/// Extension methods for differential [`Collection`]s.
+pub(crate) trait ConsolidateExt<G, D1, R>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Data,
+    D1: ExchangeData + Hashable,
+    R: Semigroup + ExchangeData,
+{
+    /// Consolidates the collection if `must_consolidate` is `true` and leaves it
+    /// untouched otherwise.
+    fn mz_consolidate_if<Tr>(&self, must_consolidate: bool, name: &str) -> Self
+    where
+        Tr: Trace + TraceReader<Key = D1, Val = (), Time = G::Timestamp, R = R> + 'static,
+        Tr::Batch: Batch,
+        Arranged<G, TraceAgent<Tr>>: ArrangementSize;
+
+    /// Consolidates the collection.
+    fn mz_consolidate<Tr>(&self, name: &str) -> Self
+    where
+        Tr: Trace + TraceReader<Key = D1, Val = (), Time = G::Timestamp, R = R> + 'static,
+        Tr::Batch: Batch,
+        Arranged<G, TraceAgent<Tr>>: ArrangementSize;
+}
+
+impl<G, D1, R> ConsolidateExt<G, D1, R> for Collection<G, D1, R>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Data,
+    D1: ExchangeData + Hashable,
+    R: Semigroup + ExchangeData,
+{
+    fn mz_consolidate_if<Tr>(&self, must_consolidate: bool, name: &str) -> Self
+    where
+        Tr: Trace + TraceReader<Key = D1, Val = (), Time = G::Timestamp, R = R> + 'static,
+        Tr::Batch: Batch,
+        Arranged<G, TraceAgent<Tr>>: ArrangementSize,
+    {
+        if must_consolidate {
+            self.mz_consolidate::<Tr>(name)
```

Comment:
Interesting, I'll defer to @vmarcos who might have more context on why this is! It might be that it only triggers within ad-hoc selects with monotonic optimizations enabled, and this might not be the case within tests.

Reply:
Yes, the line should only be executed if the feature flag `enable_monotonic_oneshot_selects` is `on` and we have a one-shot select where the optimization applies, e.g., a query with `MAX`/`MIN` or a top-k pattern. AFAIK, the tests are currently run with the feature flag turned `off`, so that we get to test our regular optimization path. The design doc #19250 has some suggestions on how to address this issue, but no solution is implemented yet at this very moment. Perhaps it would make sense to run coverage with the feature flag `on` as a validation?